File: backend/app/schemas/generation.py
# PacketArch — OT Traffic Simulation Platform
# Copyright (c) 2026 Rocky Smith <rocky.d.smith@proton.me>
# Licensed under GPL-3.0. See LICENSE at the repo root.
"""Schemas for traffic generation API."""

from datetime import datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, Field


class GenerationRequest(BaseModel):
    """Request to start traffic generation."""

    scenario_id: UUID = Field(..., description="Scenario UUID to generate traffic for")
    duration_override_ms: int | None = Field(
        None,
        description="Override scenario duration in milliseconds",
        ge=30000,
        le=3600000,
    )
    output_format: str = Field(
        "pcap",
        description="Output format (currently only 'pcap' supported)",
    )
    filename_prefix: str | None = Field(
        None,
        description="Optional prefix for output filename",
        max_length=50,
    )
    attack_playbook_id: str | None = Field(
        None,
        description=(
            "Optional attack playbook id (e.g. 'TRITON_LIKE'). When set, the "
            "attack stage packets are baked into the PCAP."
        ),
    )
    attack_config: dict[str, Any] | None = Field(
        None,
        description=(
            "Optional attack config overrides: intensity, stage_overrides, "
            "warmup_ms, start_mode."
        ),
    )
    export_attack_pcap: bool = Field(
        False,
        description=(
            "When True (and attack_playbook_id is set), the run also produces "
            "a baseline-only and an attack-only PCAP alongside the regular "
            "combined file. Ignored without a playbook."
        ),
    )
    adaptive_config: dict[str, Any] | None = Field(
        None,
        description=(
            "Optional adaptive-traffic config (AdaptiveConfig.from_dict). "
            "Enables timing drift, schedules, and vendor profiles."
        ),
    )
    cell_isolation_override: dict[str, Any] | None = Field(
        None,
        description=(
            "Optional per-run override for Purdue cell isolation. Shape: "
            "{'mode': 'off'|'conduit_gated'|'strict_northbound', "
            "'applies_to_levels': [0,1,2]}. Merged into scenario.definition "
            "before flow building so cross-cell flows are dropped before "
            "they reach the orchestrator."
        ),
    )
    topology_mode: bool = Field(
        False,
        description=(
            "Multi-sensor topology mode. When True, the run derives the L1 "
            "topology (per-zone IE3500 + core) and fans out into one PCAP per "
            "SPAN with gateway-rewritten per-segment framing. Cell isolation "
            "is forced off (cross-zone flows are routed, not dropped). Fails "
            "if the topology plan is invalid."
        ),
    )


class GenerationJobResponse(BaseModel):
    """Response with generation job details."""

    job_id: str = Field(..., description="Unique job identifier")
    scenario_id: UUID = Field(..., description="Scenario UUID")
    scenario_name: str | None = Field(None, description="Scenario name for display")
    status: str = Field(..., description="Job status (pending, running, completed, failed, cancelled)")
    progress: float = Field(0.0, description="Generation progress (0-100)", ge=0, le=100)
    total_duration_ms: int = Field(..., description="Total duration in milliseconds")
    output_path: str | None = Field(None, description="Path to output PCAP file")
    packets_generated: int = Field(0, description="Number of packets generated")
    file_size_bytes: int = Field(0, description="Output file size in bytes")
    artifacts: list[dict[str, Any]] | None = Field(
        None,
        description=(
            "PCAP files produced by the run: [{kind, filename, packets, "
            "size_bytes}]. Present for attack-export runs (combined + "
            "baseline + attack); None for single-file runs."
        ),
    )
    error_message: str | None = Field(None, description="Error message if failed")
    created_at: datetime | None = Field(None, description="Job creation timestamp")
    started_at: datetime | None = Field(None, description="Job start timestamp")
    completed_at: datetime | None = Field(None, description="Job completion timestamp")

    class Config:
        from_attributes = True


class GenerationResultResponse(BaseModel):
    """Response with generation results."""

    job_id: str = Field(..., description="Job identifier")
    scenario_id: UUID = Field(..., description="Scenario UUID")
    status: str = Field(..., description="Generation status")
    pcap_path: str | None = Field(None, description="Path to generated PCAP file")
    packets_generated: int = Field(0, description="Total packets generated")
    duration_ms: float = Field(0, description="Generation duration in milliseconds")
    file_size_bytes: int = Field(0, description="File size in bytes")
    error_message: str | None = Field(None, description="Error message if failed")


class SupportedProtocolsResponse(BaseModel):
    """Response with list of supported protocols."""

    protocols: list[str] = Field(..., description="List of supported protocol types")


class JobListResponse(BaseModel):
    """Response with list of jobs."""

    jobs: list[GenerationJobResponse] = Field(..., description="List of generation jobs")
    total: int = Field(..., description="Total number of jobs")

File: frontend/src/api/generation.ts
/*
 * PacketArch — OT Traffic Simulation Platform
 * Copyright (c) 2026 Rocky Smith <rocky.d.smith@proton.me>
 * Licensed under GPL-3.0. See LICENSE at the repo root.
 */
/**
 * API client for PCAP generation endpoints
 */

import { apiClient } from './client';

const GENERATION_PREFIX = '/api/v1/generation';

export interface GenerationArtifact {
  /** 'combined' (the regular PCAP), 'baseline' (attack removed), or 'attack' (attack only). */
  kind: 'combined' | 'baseline' | 'attack';
  filename: string;
  packets: number;
  size_bytes: number;
}

export interface GenerationJob {
  job_id: string;
  scenario_id: string;
  scenario_name?: string;
  status: 'pending' | 'running' | 'completed' | 'failed' | 'cancelled';
  progress: number;
  total_duration_ms: number;
  packets_generated: number;
  file_size_bytes: number;
  output_path?: string;
  /** Present for attack-export runs: one entry per PCAP file produced. */
  artifacts?: GenerationArtifact[] | null;
  error_message?: string;
  created_at?: string;
  started_at?: string;
  completed_at?: string;
}

export interface JobListResponse {
  jobs: GenerationJob[];
  total: number;
}

export interface StartGenerationRequest {
  scenario_id: string;
  duration_override_ms?: number;
  output_format?: string;
  filename_prefix?: string;
  /** Optional attack playbook id (e.g. 'havex_like') to bake into the PCAP. */
  attack_playbook_id?: string | null;
  /** Optional attack overrides: intensity, stage_overrides, warmup_ms, start_mode. */
  attack_config?: Record<string, unknown> | null;
  /** Optional AdaptiveConfig dict — enables timing drift and schedules. */
  adaptive_config?: Record<string, unknown> | null;
  /** Optional per-run override for Purdue cell isolation: {mode, applies_to_levels?}. */
  cell_isolation_override?: Record<string, unknown> | null;
  /** When true (with a playbook set), also emit baseline-only + attack-only PCAPs. */
  export_attack_pcap?: boolean;
}

export const generationApi = {
  /**
   * List generation jobs with optional filtering
   */
  async listJobs(params?: {
    status?: string;
    limit?: number;
    offset?: number;
  }): Promise<JobListResponse> {
    const { data } = await apiClient.get<JobListResponse>(GENERATION_PREFIX, { params });
    return data;
  },

  /**
   * Start a new PCAP generation job
   */
  async startGeneration(request: StartGenerationRequest): Promise<GenerationJob> {
    const { data } = await apiClient.post<GenerationJob>(GENERATION_PREFIX, request);
    return data;
  },

  /**
   * Get status of a specific generation job
   */
  async getJobStatus(jobId: string): Promise<GenerationJob> {
    const { data } = await apiClient.get<GenerationJob>(`${GENERATION_PREFIX}/${jobId}`);
    return data;
  },

  /**
   * Cancel a running generation job
   */
  async cancelJob(jobId: string): Promise<void> {
    await apiClient.delete(`${GENERATION_PREFIX}/${jobId}`);
  },

  /**
   * Delete a job record from history
   */
  async deleteJob(jobId: string): Promise<void> {
    await apiClient.delete(`${GENERATION_PREFIX}/${jobId}/delete`);
  },

  /**
   * Get the download URL for a completed PCAP file
   */
  getDownloadUrl(jobId: string): string {
    return `${GENERATION_PREFIX}/${jobId}/download`;
  },

  /**
   * Download a PCAP file (triggers browser download).
   *
   * @param artifact Which file to fetch for attack-export runs:
   *   'combined' (default), 'baseline', or 'attack'.
   */
  async downloadPcap(
    jobId: string,
    filename?: string,
    artifact?: 'combined' | 'baseline' | 'attack',
  ): Promise<void> {
    const response = await apiClient.get(`${GENERATION_PREFIX}/${jobId}/download`, {
      responseType: 'blob',
      params: artifact ? { artifact } : undefined,
    });

    // Create download link
    const blob = new Blob([response.data], { type: 'application/vnd.tcpdump.pcap' });
    const url = window.URL.createObjectURL(blob);
    const link = document.createElement('a');
    link.href = url;
    link.download = filename || `pcap_${jobId.substring(0, 8)}.pcap`;
    document.body.appendChild(link);
    link.click();
    document.body.removeChild(link);
    window.URL.revokeObjectURL(url);
  },

  /**
   * Get list of supported protocols
   */
  async getSupportedProtocols(): Promise<string[]> {
    const { data } = await apiClient.get<{ protocols: string[] }>(`${GENERATION_PREFIX}/protocols/supported`);
    return data.protocols;
  },
};

export default generationApi;
